metricbeat/module/consul/agent/data.go

```diff
@@ -0,0 +1,133 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type valueHelper struct {
+	Value string
+	Unit  string
+}
+
+var (
+	allowedValues = map[string]valueHelper{
+		"consul.autopilot.healthy":         {Value: "autopilot.healthy"},
+		"consul.runtime.alloc_bytes":       {Value: "runtime.alloc", Unit: ".bytes"},
+		"consul.runtime.total_gc_pause_ns": {Value: "runtime.garbage_collector.pause.total", Unit: ".ns"},
+		"consul.runtime.gc_pause_ns":       {Value: "runtime.garbage_collector.pause.current", Unit: ".ns"},
+		"consul.runtime.total_gc_runs":     {Value: "runtime.garbage_collector.runs"},
+		"consul.runtime.num_goroutines":    {Value: "runtime.goroutines"},
+		"consul.runtime.heap_objects":      {Value: "runtime.heap_objects"},
+		"consul.runtime.sys_bytes":         {Value: "runtime.sys", Unit: ".bytes"},
+		"consul.runtime.malloc_count":      {Value: "runtime.malloc_count"},
+	}
+	allowedDetailedValues = map[string]valueHelper{
+		"consul.raft.apply":      {Value: "raft.apply"},
+		"consul.raft.commitTime": {Value: "raft.commit_time", Unit: ".ms"},
+	}
+)
+
+func eventMapping(content []byte) ([]common.MapStr, error) {
+	var agent agent
+
+	if err := json.Unmarshal(content, &agent); err != nil {
+		return nil, err
+	}
+
+	labels := map[string]common.MapStr{}
+
+	for _, gauge := range agent.Gauges {
+		metricApply(labels, gauge.consulMetric, gauge.Value)
+	}
+
+	for _, point := range agent.Points {
+		metricApply(labels, point.consulMetric, point.Value)
+	}
+
+	for _, counter := range agent.Counters {
+		metricApply(labels, counter.consulMetric, consulDetailedValue(counter))
+	}
+
+	for _, sample := range agent.Samples {
+		metricApply(labels, sample.consulMetric, consulDetailedValue(sample))
+	}
+
+	data := make([]common.MapStr, 0)
+	for _, v := range labels {
+		data = append(data, v)
+	}
+
+	return data, nil
+}
+
+func metricApply(labels map[string]common.MapStr, m consulMetric, v interface{}) {
+	prettyName := prettyName(m.Name)
+	if prettyName == nil {
+		//omitting unwanted metric
+		return
+	}
+
+	labelsCombination := ""
+	for k, v := range m.Labels {
+		labelsCombination = fmt.Sprintf("%s%s%s", labelsCombination, k, v)
+	}
+
+	temp := common.MapStr{}
+	if len(m.Labels) != 0 {
+		temp.Put("labels", m.Labels)
+	}
+	key := fmt.Sprintf("%s%s", prettyName.Value, prettyName.Unit)
+
+	var value interface{}
+	switch v := v.(type) {
+	case consulDetailedValue:
+		value = v.Mean
```

Comment:
I am still not sure if mean is what we want to collect from there :thinking:

Reply:
So you are running out of time :smile:

Reply:
In general we should take the raw value.
For gauges there is no doubt, they only have a value.
For counters I'd say that `Count` would make more sense than `Mean`.
For samples I am not sure, maybe we have to take them all, but I don't see how `Mean` is better than `Max`, and we may want to have the `Count` too.

If you decide to use `Mean` please add a comment with your reasoning because I don't get it, and I'll have the same doubts if I have to revisit this code in a time.

Reply:
My _reasoning_ is that `Count` refers to the number of samples not to what you are thinking as _Counter_.
* If I fetch from Metricbeat once every 10 seconds I get one event with `Count=1`
* If I fetch from Metricbeat 10 times in 10 seconds I get 10 events with `Count=1`

```
Metric                     Count     Mean
consul.runtime.gc_pause_ns 1         112659
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
consul.runtime.gc_pause_ns 1         228525
```

Reply:
`consul.runtime.gc_pause_ns` is a `Sample`, I was not proposing _only_ `Count` for samples. There I can understand that we collect `Mean`, but maybe we should also collect other values, because why taking `Mean` and not `Max`? This is what I don't get.

Now, for `Counters`, it looks to me that `Count` is actually a raw count, in that case we could get just that. I don't understand the aggregated values there.

Reply:
We will discuss about how to support counters and samples in future PRs.